### PR TITLE
Add tests to CoreMath

### PR DESCRIPTION
**Pull request by karupayun**
Suggested by Anton, I'm implementing tests in CoreMath. In addition we
are replacing IsElementOf by ClosedInterval::Contains().

Bug: http://b/229089446

### CODE REVIEW

File: src/OrbitGl/CoreMath.h [hunk 1]
```diff
@@ -36,18 +36,20 @@ struct ClosedInterval {
     return this->min <= closed_interval.max && this->max >= closed_interval.min;
   }
 
+  [[nodiscard]] bool Contains(float value) { return this->min <= value && this->max >= value; }
+
+  bool operator==(ClosedInterval const& interval) const {
```
Comment (antonrohr) [inline src/OrbitGl/CoreMath.h:41]: `const ClosedInterval&` instead of `ClosedInterval const&`?

File: src/OrbitGl/CoreMathTest.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  EXPECT_EQ(kFakeClosedInterval, ClosedInterval::FromValues(3, 0));
+}
+
+TEST(ClosedInterval, Intersects) {
+  EXPECT_TRUE((ClosedInterval{0, 3}.Intersects(ClosedInterval{1, 2})));
+  EXPECT_TRUE((ClosedInterval{1, 2}.Intersects(ClosedInterval{0, 3})));
+  EXPECT_TRUE((ClosedInterval{0, 1}.Intersects(ClosedInterval{1, 2})));
+  EXPECT_TRUE((ClosedInterval{1, 2}.Intersects(ClosedInterval{0, 1})));
+  EXPECT_FALSE((ClosedInterval{0, 1}.Intersects(ClosedInterval{2, 3})));
+  EXPECT_FALSE((ClosedInterval{2, 3}.Intersects(ClosedInterval{0, 1})));
```
Comment (antonrohr) [inline src/OrbitGl/CoreMathTest.cpp:24]: I think there are extra `( )` that are not needed

**Review by antonrohr — APPROVED** [review]
Very nice, I like it a lot!

File: src/OrbitGl/CoreMathTest.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  EXPECT_TRUE(kFakeClosedInterval.Contains(0));
+  EXPECT_TRUE(kFakeClosedInterval.Contains(1));
+  EXPECT_TRUE(kFakeClosedInterval.Contains(3));
+  EXPECT_FALSE(kFakeClosedInterval.Contains(4));
+}
+
+TEST(CoreMath, IsInsideRectangle) {
+  const Vec2 kFakeRectangleTopLeft{5, 5};
+  const Vec2 kFakeRectangleSize{2, 2};
+  EXPECT_FALSE(IsInsideRectangle({4, 4}, kFakeRectangleTopLeft, kFakeRectangleSize));
```
Comment (vickyliu-go4it) [inline src/OrbitGl/CoreMathTest.cpp:38]: Do we really need so many test cases 😂 ?

**Review by vickyliu-go4it — APPROVED** [review]
LGTM!

File: src/OrbitGl/CoreMathTest.cpp (hunk 2 above)
Comment (karupayun) [reply src/OrbitGl/CoreMathTest.cpp:24]: I thought the same, but the compiler doesn't agree with us:

https://screenshot/AFodPCSuFdDJAuv

File: src/OrbitGl/CoreMathTest.cpp (hunk 3 above)
Comment (karupayun) [reply src/OrbitGl/CoreMathTest.cpp:38]: For the FALSE we have one case in the top left, one in the bottom right, one at the top, one at the bottom, one at the left and one at the right.

For the TRUE, we have one per side + one in the center.

File: src/OrbitGl/CoreMathTest.cpp (hunk 2 above)
Comment (antonrohr) [reply src/OrbitGl/CoreMathTest.cpp:24]: Interesting!

File: src/OrbitGl/CoreMathTest.cpp (hunk 3 above)
Comment (vickyliu-go4it) [reply src/OrbitGl/CoreMathTest.cpp:38]: I'm fine to keep it as now.

What I mean is, here our rectangle is a closure. So from my mind, I can only see three cases:
- nodes on the boundary
- nodes inside the closure
- nodes out side the closure

For the TRUE case,  you test four corners and the center node. But actually there are also some other nodes on the rectangle boundary. If we list it in that case, we can also test the center point of the left edge, and the center point of the top edge of the rectangle.

Similarly, for the FALSE case, we test cases for points outside the closure. Then in addition to the 6 cases you listed above, there are also bottom left and top right.
But this is just my personal preference 😂

File: src/OrbitGl/CoreMath.h [hunk 4]
```diff
@@ -36,18 +36,20 @@ struct ClosedInterval {
     return this->min <= closed_interval.max && this->max >= closed_interval.min;
   }
 
+  [[nodiscard]] bool Contains(float value) { return this->min <= value && this->max >= value; }
```
Comment (beckerhe) [inline src/OrbitGl/CoreMath.h:39]: const?

File: src/OrbitGl/CoreMath.h [hunk 5]
```diff
@@ -36,18 +36,20 @@ struct ClosedInterval {
...
+
+  bool operator==(const ClosedInterval& interval) const {
```
Comment (beckerhe) [inline src/OrbitGl/CoreMath.h:41]: Can you use the hidden friend idiom?

Comment (beckerhe) [reply src/OrbitGl/CoreMath.h:41]: Nit: And nodiscard.

Comment (beckerhe) [reply src/OrbitGl/CoreMath.h:41]: And please also define operator!= by calling operator==. They should always come in pairs.

File: src/OrbitGl/CoreMathTest.cpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <gtest/gtest.h>
+
+#include "CoreMath.h"
+
+namespace orbit_gl {
+
+ClosedInterval kFakeClosedInterval{0, 3};
```
Comment (beckerhe) [inline src/OrbitGl/CoreMathTest.cpp:11]: const

File: src/OrbitGl/CoreMathTest.cpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+TEST(ClosedInterval, FromValues) {
+  EXPECT_EQ(kFakeClosedInterval, ClosedInterval::FromValues(0, 3));
+  EXPECT_EQ(kFakeClosedInterval, ClosedInterval::FromValues(3, 0));
+}
```
Comment (beckerhe) [inline src/OrbitGl/CoreMathTest.cpp:16]: This makes not a lot of sense to me. If you don't force your users to use `FromValues` it's better not to have it IMHO.

File: src/OrbitGl/CoreMathTest.cpp [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+TEST(ClosedInterval, Intersects) {
```
Comment (beckerhe) [inline src/OrbitGl/CoreMathTest.cpp:18]: What about also checking single value intervals, like `ClosedInterval::FromValues(0,0)`?

File: src/OrbitGl/CoreMathTest.cpp (hunk 6 above)
Comment (beckerhe) [reply src/OrbitGl/CoreMathTest.cpp:11]: And why `Fake`? It's not a fake, is it?

**Review by beckerhe — APPROVED** [review]
Otherwise LGTM

File: src/OrbitGl/CoreMathTest.cpp (hunk 7 above)
Comment (karupayun) [reply src/OrbitGl/CoreMathTest.cpp:16]: Yes, I thought about that. I will revert it.
